tests/h/subscribers_test.py

```diff
@@ -190,3 +171,28 @@ def transaction_manager(self, pyramid_request):
             TransactionManager, instance=True, spec_set=True
         )
         return pyramid_request.tm
+
+
+@pytest.fixture(autouse=True)
+def storage(patch):
+    return patch("h.subscribers.storage")
+
+
+@pytest.fixture(autouse=True)
+def reply(patch):
+    return patch("h.subscribers.reply")
+
+
+@pytest.fixture(autouse=True)
+def mailer(patch):
+    return patch("h.subscribers.mailer")
+
+
+@pytest.fixture(autouse=True)
+def emails(patch):
+    return patch("h.subscribers.emails")
+
+
+@pytest.fixture(autouse=True)
+def report_exception(patch):
+    return patch("h.subscribers.report_exception")
```

Comment:
In general I understand that we put these here as you like to compare them with the imports, but I think in this case something is lost.

Normally things in the same file are going to want most of the same patches etc, but this is a counter-example where the things in the `subscribers.py` file really have nothing much to do with each other apart from being in a particular style.

This isn't a massive problem here as everything is relatively small, but should we choose to ship out the subscribers into a separate file, it's quite nice to lift and shift the test class and have everything that's relevant to it move. It also means that these patches are applied a number of times where they don't really need to be.

Maybe all of this is just to say we'd be better off splitting these into separate files instead? If we really have two or more functionally distinct blobs of code in the same place, maybe that's the real issue?

Reply:
Yeah I think splitting them into separate files may be the right way.

In the past I've found that when a module imports something that should be patched, such as `subscribers.py` doing `from h_pyramid_sentry import report_exception` in this case, it's a good idea for the `report_exception` patch fixture to be a module-scoped fixture in the test file `subscribers_test.py` with `autouse=True`, like it is here, rather than a class-scoped fixture. The reason is that the imported name (`report_exception`) is a module-scoped name in `subscribers.py` so code anywhere in `subscribers.py` could use it. It may be that, right now, only one `subscribers.py` class or function does use `report_exception`. But it's all too easy to modify `subscribers.py` adding a `report_exception` usage to another part of the file and forget to also add the patch fixture to the test class for _that_ part of the file, and end up with tests calling the real `h_pyramid_sentry.report_exception()`.

This is one of those things that I've seen again and again in h and had to debug fix. That's why I introduced this "patch fixtures are autoused module-scoped fixtures" pattern, as it eliminates this class of errors.

The ability to compare the list of fixtures with the imports is also nice.